### PR TITLE
Prac 09 feedback

### PR DESCRIPTION
**Pull request by Yangzijing16**
please review my practical 9 @yufei-leo

### CODE REVIEW

File: prac_09/band.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        """
+        self.name = name
+        self.musicians = musicians
+
+    def __str__(self) -> str:
+        """Return a string representation of the band.
+
+        Returns:
+            str: Band name and musician details.
+        """
```
Comment (Yangzijing16) [inline prac_09/band.py:24]: `if not self.musicians:
            return f"{self.name} (no musicians)"
        return f"{self.name} ({', '.join(str(musician) for musician in self.musicians)})"`

File: prac_09/silver_service_taxi.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self, name: str, fuel: float, fanciness: float):
+        """Initialise a SilverServiceTaxi with name, fuel, and fanciness.
+
+        Args:
+            name (str): The name of the taxi.
+            fuel (float): The amount of fuel.
+            fanciness (float): Multiplier for price per km.
+        """
+        super().__init__(name, fuel)
```
Comment (Yangzijing16) [inline prac_09/silver_service_taxi.py:20]: self.fanciness = fancinesss

File: prac_09/silver_service_taxi.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            str: Description including parent details and flagfall.
+        """
+        return f"{super().__str__()} plus flagfall of ${FLAGFALL:.2f}"
+
+    def get_fare(self) -> float:
+        """Calculate the fare including flagfall, rounded to nearest 10c.
+
+        Returns:
+            float: The total fare for the trip.
+        """
```
Comment (Yangzijing16) [inline prac_09/silver_service_taxi.py:36]: total_fare = super().get_fare() + self.flagball

File: prac_09/silver_service_taxi_test.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+from taxi import PRICE_PER_KM
+from silver_service_taxi import FLAGFALL
+
+
+def main():
+    """Test the SilverServiceTaxi class with a sample trip."""
+    taxi = SilverServiceTaxi("Hummer", 200, 2)
+
+    # Drive 18 km
+    taxi.drive(18)
```
Comment (Yangzijing16) [inline prac_09/silver_service_taxi_test.py:13]: distance = 18

File: prac_09/unreliable_car.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+        Args:
+            distance (float): The distance to attempt to drive.
+
+        Returns:
+            float: The actual distance driven (0 if unreliable).
+        """
+        if random.uniform(0, 100) < self.reliability:
+            return super().drive(distance)
+        return 0
```
Comment (Yangzijing16) [inline prac_09/unreliable_car.py:32]: ` def _is_reliable(self) -> bool:
        """Determine whether the car drives based on its reliability."""
        return random.uniform(0, 100) < self.reliability

    def __str__(self) -> str:
        """
        Return a string representation of the UnreliableCar.

        Returns:
            str: Description including name, fuel, and reliability.
        """
        return f"{super().__str__()} (Reliability: {self.reliability:.1f}%)"`

**Comment by yufei-leo** [comment]
This code demonstrates solid object-oriented design with clear inheritance structures (Taxi/SilverServiceTaxi from Car, UnreliableCar from Car) and good separation of concerns across modules. The implementation includes proper type hints, docstrings, and test files for each main class, showing attention to maintainability. However, there are opportunities for improvement in input validation (especially for negative distances and reliability percentages), more consistent documentation formats, and more robust error handling. The testing could be enhanced with additional edge cases and tighter assertion ranges, particularly for the UnreliableCar tests. Some code quality issues like typos ("fancinesss") and missing str overrides should be addressed. The architecture would benefit from centralizing constants and making the TaxiFleet more generic. Overall, the foundation is well-structured but would become more production-ready with additional validation, polishing, and comprehensive test coverage.
